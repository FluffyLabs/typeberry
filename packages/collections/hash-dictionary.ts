import type { OpaqueHash } from "@typeberry/hash";

/** A map which uses hashes as keys. */
export class HashDictionary<K extends OpaqueHash, V> {
  // TODO [ToDr] [crit] We can't use `TrieHash` directly in the map,
  // because of the way it's being compared. Hence having `string` here.
  // This has to be benchmarked and re-written to a custom map most likely.
  protected readonly map = new Map<string, [K, V]>();

  /** Return number of items in the dictionary. */
  get size(): number {
    return this.map.size;
  }

  /** Return true if the key is present in the dictionary. */
  has(key: K): boolean {
    return this.map.has(key.toString());
  }

  /** Get the value under given key or `null` if the value is not present. */
  get(key: K): V | undefined {
    return this.map.get(key.toString())?.[1];
  }

  /** Insert/overwrite the value at given key. */
  set(key: K, value: V) {
    this.map.set(key.toString(), [key, value]);
  }

  /** Remove the key and any value from the dictionary. */
  delete(key: K) {
    this.map.delete(key.toString());
  }

  /** it allows to use HashDictionary in for-of loop */
  *[Symbol.iterator]() {
    for (const value of this.map.values()) {
      yield value;
    }
  }
}
