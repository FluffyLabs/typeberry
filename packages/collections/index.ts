export * from "./hash-dictionary";
export * from "./sized-array";
export * from "./sorted-array";
