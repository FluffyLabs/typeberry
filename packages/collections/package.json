{
  "name": "@typeberry/collections",
  "version": "0.0.1",
  "main": "index.ts",
  "scripts": {
    "test": "node --test --require ts-node/register $(find . -type f -o -name '*.test.ts' | tr '\\n' ' ')"
  },
  "author": "Fluffy Labs",
  "license": "MPL-2.0",
  "description": "Known-size collection types."
}
