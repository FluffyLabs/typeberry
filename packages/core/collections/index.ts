export * from "./hash-dictionary.js";
export * from "./hash-set.js";
export * from "./multi-map.js";
export * from "./sized-array.js";
export * from "./sorted-array.js";
export * from "./sorted-set.js";
export * from "./truncated-hash-dictionary.js";
