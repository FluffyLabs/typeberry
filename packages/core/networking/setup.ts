import QuicLogger, { formatting, LogLevel, StreamHandler } from "@matrixai/logger";
import { events, QUICClient, type QUICConnection, QUICServer, QUICSocket } from "@matrixai/quic";
import { BytesBlob } from "@typeberry/bytes";
import type { Ed25519Pair } from "@typeberry/crypto/ed25519.js";
import { Level, Logger } from "@typeberry/logger";
import {
  altNameRaw,
  certToPEM,
  ed25519AsJsonWebKeyPair,
  generateCertificate,
  type PeerInfo,
  privateKeyToPEM,
} from "./certificate.js";
import { getQuicClientCrypto, getQuicServerCrypto } from "./crypto.js";
import type { DialOptions } from "./network.js";
import { peerVerification } from "./peer-verification.js";
import { type PeerAddress, PeersManagement } from "./peers.js";
import { QuicNetwork } from "./quic-network.js";
import { QuicPeer } from "./quic-peer.js";
import { addEventListener } from "./quic-utils.js";

const logger = Logger.new(import.meta.filename, "net");

/** Networking server part options. */
export type Options = {
  /** Peer's ed25519key. */
  key: Ed25519Pair;
  /** Host interface to bound to. */
  host: string;
  /** Port to listen on. Use `0` for random. */
  port: number;
  /** Supported ALPNs (protocols) both for the inbound and outbound connections. */
  protocols: string[];
};

enum CloseReason {
  PeerIdMismatch = 0,
  DuplicateConnection = 1,
  ConnectionFromOurself = 2,
}

export class Quic {
  /** Setup QUIC socket and start listening for connections. */
  static async setup({ host, port, protocols, key }: Options): Promise<QuicNetwork> {
    const quicLoggerLvl = logger.getLevel() > Level.TRACE ? LogLevel.WARN : LogLevel.DEBUG;
    const quicLogger = new QuicLogger("quic", quicLoggerLvl, [
      new StreamHandler(formatting.format`${formatting.level}:${formatting.keys}:${formatting.msg}`),
    ]);

    // Load keypair
    const keyPair = ed25519AsJsonWebKeyPair(key);
    const privKeyPEM = await privateKeyToPEM(keyPair);
    const cert = await generateCertificate({
      certId: BytesBlob.blobFromString("QUIC Networking"),
      subjectKeyPair: keyPair,
      issuerKeyPair: keyPair,
    });

    const lastConnectedPeer = peerVerification();

    // QUICConfig
    const config = {
      keepAliveIntervalTime: 3000,
      maxIdleTimeout: 6000,
      applicationProtos: protocols,
      cert: certToPEM(cert),
      key: privKeyPEM,
      verifyPeer: true,
      verifyCallback: lastConnectedPeer.verifyCallback,
    };

    logger.info`🆔 Peer id: ** ${altNameRaw(key.pubKey)}@${host}:${port} ** (pubkey: ${key.pubKey})`;
    // Shared injected UDP socket
    const socket = new QUICSocket({
      logger: quicLogger.getChild("socket"),
    });

    // Start server on the socket.
    const server = new QUICServer({
      socket,
      config,
      crypto: getQuicServerCrypto(key),
      logger: quicLogger.getChild("server"),
    });

    // peer management
    const peers = new PeersManagement<QuicPeer>();

    // basic error handling
    addEventListener(server, events.EventQUICServerError, (error) => logger.error`🛜  Server error: ${error}`);
    addEventListener(server, events.EventQUICServerClose, (ev) => logger.error`🛜  Server stopped: ${ev}`);

    // handling incoming session
    addEventListener(server, events.EventQUICServerConnection, async (ev) => {
      const conn = ev.detail;
      if (lastConnectedPeer.info === null) {
        await conn.stop();
        return;
      }

      if (lastConnectedPeer.info.key.isEqualTo(key.pubKey)) {
        logger.log`🛜 Rejecting connection from ourself from ${conn.remoteHost}:${conn.remotePort}`;
        await conn.stop({ isApp: true, errorCode: CloseReason.ConnectionFromOurself });
        return;
      }

      if (peers.isConnected(lastConnectedPeer.info.id)) {
        logger.log`🛜 Rejecting duplicate connection with peer ${lastConnectedPeer.info.id} from ${conn.remoteHost}:${conn.remotePort}`;
        await conn.stop({ isApp: true, errorCode: CloseReason.DuplicateConnection });
        return;
      }

      logger.log`🛜 Server handshake with ${conn.remoteHost}:${conn.remotePort}`;
      newPeer(conn, lastConnectedPeer.info);
      lastConnectedPeer.info = null;
      await conn.start();
    });

    // connecting to a peer
    async function dial(peer: PeerAddress, options: DialOptions): Promise<QuicPeer> {
      return doDial();

      async function doDial() {
        const peerDetails = peerVerification();
        const clientLater = QUICClient.createQUICClient(
          {
            socket: socket,
            host: peer.host,
            port: peer.port,
            crypto: getQuicClientCrypto(),
            config: {
              ...config,
              verifyCallback: peerDetails.verifyCallback,
            },
            logger: quicLogger.getChild("client"),
          },
          {
            signal: options.signal,
          },
        );
        const client = await clientLater;

        if (peerDetails.info === null) {
          await client.destroy({ isApp: true, errorCode: CloseReason.PeerIdMismatch });
          throw new Error("Client connected, but there is no peer details!");
        }

        if (options.verifyName !== undefined && options.verifyName !== peerDetails.info.id) {
          await client.destroy({ isApp: true, errorCode: CloseReason.PeerIdMismatch });
          throw new Error(
            `Client connected, but the id didn't match. Expected: ${options.verifyName}, got: ${peerDetails.info.id}`,
          );
        }

        addEventListener(client, events.EventQUICClientClose, () => {
          logger.log`⚰️ Client connection closed.`;
        });

        addEventListener(client, events.EventQUICClientError, (error) => {
          logger.error`🔴 Client error: ${error.detail}`;
        });

        logger.log`🤝 Client handshake with: ${peer.host}:${peer.port}`;
        return newPeer(client.connection, peerDetails.info);
      }
    }

    function newPeer(conn: QUICConnection, peerInfo: PeerInfo) {
      const peer = new QuicPeer(conn, peerInfo);
      addEventListener(peer.conn, events.EventQUICConnectionClose, () => peers.peerDisconnected(peer));
      peers.peerConnected(peer);
      return peer;
    }

    return new QuicNetwork(socket, server, dial, peers, { host, port });
  }
}
