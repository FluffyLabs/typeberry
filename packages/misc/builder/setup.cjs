const fs = require("node:fs");
const childProcess = require("node:child_process");

const packageToBuild = process.env.PACKAGE_NAME;
let outDir = process.env.PACKAGE_OUT;

const isRelease = process.env.IS_RELEASE !== undefined;

if (!packageToBuild) {
  throw new Error(`Missing 'PACKAGE_NAME' environment variable.`);
}

if (!outDir) {
  const parts = packageToBuild.split("/");
  outDir = parts[parts.length - 1];
}

const DIST = `${__dirname}/../../../dist/${outDir}`;
const inputFile = `${__dirname}/pkg.ts`;

const data = `export * from "${packageToBuild}";`;
fs.writeFileSync(inputFile, data);

const commitHashResult = childProcess.execSync("git rev-parse --short HEAD");
const originalPackageJson = require(`${packageToBuild}/package.json`);
// generate package.json
const packageJson = JSON.stringify(
  {
    name: packageToBuild,
    version: isRelease
      ? originalPackageJson.version
      : `${originalPackageJson.version}-${commitHashResult.toString("utf8").trim()}`,
    main: "index.js",
    author: originalPackageJson.author,
    license: originalPackageJson.license,
    sideEffects: false,
    exports: {
      ".": {
        import: "./index.js",
        require: "./index.cjs",
      },
    },
  },
  null,
  2,
);
fs.mkdirSync(DIST, { recursive: true });
fs.writeFileSync(`${DIST}/package.json`, packageJson);

fs.writeFileSync(
  `${DIST}/.npmignore`,
  `
bin/**
packages/**
`,
);

module.exports = {
  inputFile,
  esmOutFile: `${DIST}/index.js`,
  cjsOutFile: `${DIST}/index.cjs`,
  typesInput: `${DIST}/packages/misc/builder/pkg.d.ts`,
};
