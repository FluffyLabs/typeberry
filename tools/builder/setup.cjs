const fs = require("node:fs");
const childProcess = require("node:child_process");

const packageToBuild = process.env.PACKAGE_NAME;
let outDir = process.env.PACKAGE_OUT;

if (!packageToBuild) {
  throw new Error(`Missing 'PACKAGE_NAME' environment variable.`);
}

if (!outDir) {
  const parts = packageToBuild.split("/");
  outDir = parts[parts.length - 1];
}

const data = `export * from "${packageToBuild}";`;
fs.writeFileSync(`${__dirname}/pkg.ts`, data);

const DIST = `${__dirname}/../../dist/${outDir}`;

const commitHashResult = childProcess.execSync("git rev-parse --short HEAD");
const originalPackageJson = require(`${packageToBuild}/package.json`);
// generate package.json
const packageJson = JSON.stringify(
  {
    name: packageToBuild,
    version: `${originalPackageJson.version}-${commitHashResult.toString("utf8").trim()}`,
    main: "index.js",
    author: originalPackageJson.author,
    license: originalPackageJson.license,
  },
  null,
  2,
);
fs.mkdirSync(DIST, { recursive: true });
fs.writeFileSync(`${DIST}/package.json`, packageJson);

fs.writeFileSync(
  `${DIST}/.npmignore`,
  `tools/**
packages/**
`,
);

module.exports = {
  outFile: `${DIST}/index.js`,
  typesInput: `${DIST}/tools/builder/pkg.d.ts`,
};
