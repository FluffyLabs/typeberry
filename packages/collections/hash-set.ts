import type { OpaqueHash } from "@typeberry/hash";
import { HashDictionary } from "./hash-dictionary";

/** A set specialized for storing hashes. */
export class HashSet<V extends OpaqueHash> {
  private readonly map = new HashDictionary<V, boolean>();

  /** Return number of items in the set. */
  get size(): number {
    return this.map.size;
  }

  /** Insert given hash to the set. */
  insert(value: V) {
    return this.map.set(value, true);
  }

  /** Check if given hash is in the set. */
  has(value: V) {
    return this.map.has(value);
  }

  /**
   * Remove value from set.
   *
   * Returns `true` if element existed in the set and was removed, `false` otherwise.
   */
  delete(value: V) {
    return this.map.delete(value);
  }

  /** it allows to use HashSet in for-of loop */
  *[Symbol.iterator]() {
    for (const value of this.map) {
      yield value[0];
    }
  }
}
