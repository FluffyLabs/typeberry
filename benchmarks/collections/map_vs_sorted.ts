import { add, complete, configure, cycle, save, suite } from "@typeberry/benchmark/setup";
import { Ordering, SortedArray } from "@typeberry/collections";

const READS = 100;
const keys = ["xyz", "abc", "123", "def", "Abb"];
const converted = keys.map((key) => ({ key }));

module.exports = () =>
  suite(
    "Map vs SortedArray for small element count",

    add("Map", () => {
      const map = new Map();
      map.set(keys[0], { key: keys[0], value: true });
      map.set(keys[1], { key: keys[1], value: false });
      map.set(keys[2], { key: keys[2], value: true });

      return () => {
        for (let k = 0; k < READS; k += 1) {
          for (const field of converted) {
            const v = map.get(field.key);
            if (v) {
              dataCmp(v, field) === Ordering.Equal;
            }
          }
        }
      };
    }),

    add("Map-array", () => {
      const map = new Map();
      map.set(0, { key: keys[0], value: true });
      map.set(1, { key: keys[1], value: false });
      map.set(2, { key: keys[2], value: true });
      const len = map.size;
      return () => {
        for (let k = 0; k < READS; k += 1) {
          for (const field of converted) {
            for (let i = 0; i < len; i += 1) {
              const v = map.get(i);
              if (dataCmp(v, field) === Ordering.Equal) {
                break;
              }
            }
          }
        }
      };
    }),

    add("Array", () => {
      const map: Data[] = [];
      map.push({ key: keys[0], value: true });
      map.push({ key: keys[1], value: false });
      map.push({ key: keys[2], value: true });

      return () => {
        for (let k = 0; k < READS; k += 1) {
          for (const field of converted) {
            map.findIndex((v) => dataCmp(v, field) === Ordering.Equal);
          }
        }
      };
    }),

    add("SortedArray", () => {
      const map = new SortedArray<Data>(dataCmp);
      map.insert({ key: keys[0], value: true });
      map.insert({ key: keys[1], value: false });
      map.insert({ key: keys[2], value: true });

      return () => {
        for (let k = 0; k < READS; k += 1) {
          for (const field of converted) {
            map.findIndex(field);
          }
        }
      };
    }),

    cycle(),
    complete(),
    configure({}),
    ...save(__filename),
  );

type Data = { key: string; value?: boolean };
function dataCmp(a: Data, b: Data) {
  if (a.key < b.key) {
    return Ordering.Less;
  }

  if (a.key > b.key) {
    return Ordering.Greater;
  }

  return Ordering.Equal;
}

if (require.main === module) {
  module.exports();
}
