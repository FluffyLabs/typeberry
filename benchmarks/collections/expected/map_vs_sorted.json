{
  "name": "Map vs SortedArray for small element count",
  "date": "2024-10-23T20:32:16.451Z",
  "version": null,
  "results": null,
  "fastest": {
    "name": "Map",
    "index": 0
  },
  "slowest": {
    "name": "SortedArray",
    "index": 2
  }
}
